File: 백준/Gold/7569. 토마토/토마토.js
const input = require('fs')
  .readFileSync(0)
  .toString()
  .trim()
  .split('\n');

const [M, N, H] = input.shift().split(' ').map(Number);

const dx = [-1, 1, 0, 0, 0, 0];
const dy = [0, 0, -1, 1, 0, 0];
const dz = [0, 0, 0, 0, -1, 1];
const queue = [];

const boxes = Array.from(
  Array(H),
  () => Array.from(N),
  () => Array.from(Array(M).fill(0))
);

for (let i = 0; i < H; i += 1) {
  for (let j = 0; j < N; j += 1) {
    boxes[i][j] = input.shift().split(' ').map(Number);
  }
}

let unripeTomatoes = 0;
for (let z = 0; z < H; z += 1) {
  for (let x = 0; x < N; x += 1) {
    for (let y = 0; y < M; y += 1) {
      if (boxes[z][x][y] === 0) unripeTomatoes += 1;
      if (boxes[z][x][y] === 1) queue.push([z, x, y, 0]);
    }
  }
}

let index = 0;
let answer = 0;

while (queue.length > index) {
  const [z, x, y, days] = queue[index++];

  for (let i = 0; i < 6; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];
    const nz = z + dz[i];

    if (
      nz >= 0 &&
      nz < H &&
      nx >= 0 &&
      nx < N &&
      ny >= 0 &&
      ny < M &&
      !boxes[nz][nx][ny]
    ) {
      boxes[nz][nx][ny] = 1;
      queue.push([nz, nx, ny, days + 1]);
      unripeTomatoes -= 1;
    }
  }

  answer = days;
}

console.log(unripeTomatoes ? -1 : answer);

File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque

def bfs():
    global queue

    while queue:
        x, y, z = queue.popleft()
        for i in range(6):
            nx = x + dx[i]
            ny = y + dy[i]
            nz = z + dz[i]

            if 0 <= nx < N and 0 <= ny < M and 0 <= nz < H:
                if boxes[nz][nx][ny] == 0:
                    boxes[nz][nx][ny] = boxes[z][x][y] + 1
                    queue.append((nx, ny, nz))

M, N, H = map(int, input().split())
boxes = [[list(map(int, input().split())) for _ in range(N)] for _ in range(H)]

# 위 아래 왼쪽 오른쪽 앞 뒤
dx = [0, 0, 1, -1, 0, 0]
dy = [1, -1, 0, 0, 0, 0]
dz = [0, 0, 0, 0, 1, -1]

queue = deque()

for height in range(H):
    for row in range(N):
        for col in range(M):
            if boxes[height][row][col] == 1:
                queue.append((row, col, height))

bfs()
answer = 0
uncompleted = False

for height in range(H):
    for row in range(N):
        for col in range(M):
            if boxes[height][row][col] == 0:
                uncompleted = True
            answer = max(answer, boxes[height][row][col])

if uncompleted:
    print(-1)
else:
    print(answer - 1)